### DISCUSSION TITLE
401 Unauthorized error using JS-SDK

### DISCUSSION THREAD

**Discussion by DYSLEVIUM** [deepgram/community#590] [General help] [answered]
I am currently working on a React Native application where I am integrating Deepgram for audio transcription. I have successfully implemented the transcription feature using the Deepgram Python backend, but I am facing challenges when attempting to integrate it into my React Native application.

I am consistently receiving a 401 Unauthorized error even though I have confirmed that the API key is correct. Also, I am encountering an error related to the base URL when making a connection to the Deepgram API. The error suggests that the base URL provided does not begin with 'http,' 'https,' 'ws,' or 'wss' and defaults to 'https' as standard.

Code: 
```typescript
import { LiveTranscriptionEvents, createClient } from '@deepgram/sdk';
import { useCallback, useEffect, useMemo, useRef } from 'react';

const useDeepgram = () => {
  const API_KEY = 'API_KEY';
  const client = useMemo(
    () => createClient(API_KEY),
    []
  );

  const keepAlive = useRef<NodeJS.Timeout>();
  const setupDeepgram = useCallback(() => {
    const deepgram = client.listen.live({
      language: 'en-US',
      detect_language: true,
      interim_results: true,
      punctuate: true,
      model: 'nova-2-general',
      smart_format: true,
    });

    if (keepAlive.current) clearInterval(keepAlive.current);
    keepAlive.current = setInterval(() => {
      console.log('deepgram: keepalive');
      deepgram.keepAlive();
    }, 10 * 1000);

    deepgram.addListener(LiveTranscriptionEvents.Open, async () => {
      console.log('deepgram: connected');

      deepgram.addListener(LiveTranscriptionEvents.Transcript, (data) => {
        console.log('deepgram: packet received');
        console.log('deepgram: transcript received', data);
      });

      deepgram.addListener(LiveTranscriptionEvents.Close, async () => {
        console.log('deepgram: disconnected');
        clearInterval(keepAlive.current);
        deepgram.finish();
      });

      deepgram.addListener(LiveTranscriptionEvents.Error, async (error) => {
        console.log('deepgram: error received', error);
      });

      deepgram.addListener(LiveTranscriptionEvents.Warning, async (warning) => {
        console.log('deepgram: warning received', warning);
      });

      deepgram.addListener(LiveTranscriptionEvents.Metadata, (data) => {
        console.log('deepgram: packet received');
        console.log('deepgram: metadata received', data);
      });
    });

    return deepgram;
  }, [client]);

  const deepgram = useMemo(setupDeepgram, [setupDeepgram]);

  useEffect(() => {
    return () => {
      deepgram.finish();
      deepgram.removeAllListeners();
    };
  }, [deepgram]);

  return deepgram;
};

export default useDeepgram;
```

Error with debug logs: 
```javascript
LOG  deepgram connection: {
  "_events": {},
  "_eventsCount": 1,
  "key": "API_KEY",
  "options": {
    "global": {
      "url": "api.deepgram.com"
    },
    "fetch": {
      "headers": {
        "Content-Type": "application/json",
        "X-Client-Info": "@deepgram/sdk; browser; v3.1.6",
        "User-Agent": "@deepgram/sdk/3.1.6"
      }
    }
  },
  "baseUrl": "wss://api.deepgram.com/",
  "transcriptionOptions": {
    "language": "en-US",
    "detect_language": true,
    "interim_results": true,
    "punctuate": true,
    "model": "nova-2-general",
    "smart_format": true
  },
  "_socket": {
    "CONNECTING": 0,
    "OPEN": 1,
    "CLOSING": 2,
    "CLOSED": 3,
    "readyState": 0,
    "url": "wss://api.deepgram.com/v1/listen?language=en-US&detect_language=true&interim_results=true&punctuate=true&model=nova-2-general&smart_format=true",
    "_eventEmitter": {},
    "_socketId": 3,
    "_subscriptions": [
      {},
      {},
      {},
      {}
    ]
  }
}

Error: Unhandled error. (Expected HTTP 101 response but was '401 Unauthorized')

LOG  Error has been symbolicated
Error: The base URL provided does not begin with http, https, ws, or wss and will default to https as standard.
    at constructor (/Users/pb/te/react-native-sdk/lv/node_modules/@deepgram/sdk/dist/main/packages/AbstractClient.js:35:24)
    at <global> (/Users/pb/te/react-native-sdk/lv/node_modules/@deepgram/sdk/dist/main/DeepgramClient.js:1:13)
    at DeepgramClient (/Users/pb/te/react-native-sdk/lv/node_modules/@deepgram/sdk/dist/main/DeepgramClient.js:15:20)
    at createClient (/Users/pb/te/react-native-sdk/lv/node_modules/@deepgram/sdk/dist/main/index.js:40:39)
    at useMemo$argument_0 (/Users/pb/te/react-native-sdk/lv/src/hooks/useDeepgram.ts:7:22)
    at useMemo (/Users/pb/te/react-native-sdk/lv/node_modules/react/cjs/react.development.js:1650:27)
    at useDeepgram (/Users/pb/te/react-native-sdk/lv/src/hooks/useDeepgram.ts:6:24)
    at TranscriptPage (/Users/pb/te/react-native-sdk/lv/src/TranscriptPage.tsx:14:30)
 LOG  error: [Error: Unhandled error. (Expected HTTP 101 response but was '401 Unauthorized')] Error: Unhandled error. (Expected HTTP 101 response but was '401 Unauthorized')
    at emit (http://localhost:8081/index.bundle//&platform=android&dev=true&minify=false&app=com.app.lv&modulesOnly=false&runModule=true:139765:26)
    at anonymous (http://localhost:8081/index.bundle//&platform=android&dev=true&minify=false&app=com.app.lv&modulesOnly=false&runModule=true:139301:19)
    at call (native)
    at dispatchEvent (http://localhost:8081/index.bundle//&platform=android&dev=true&minify=false&app=com.app.lv&modulesOnly=false&runModule=true:37099:31)
    at anonymous (http://localhost:8081/index.bundle//&platform=android&dev=true&minify=false&app=com.app.lv&modulesOnly=false&runModule=true:38238:31)
    at apply (native)
    at emit (http://localhost:8081/index.bundle//&platform=android&dev=true&minify=false&app=com.app.lv&modulesOnly=false&runModule=true:2503:40)
    at apply (native)
    at __callFunction (http://localhost:8081/index.bundle//&platform=android&dev=true&minify=false&app=com.app.lv&modulesOnly=false&runModule=true:3343:36)
    at anonymous (http://localhost:8081/index.bundle//&platform=android&dev=true&minify=false&app=com.app.lv&modulesOnly=false&runModule=true:3064:31)
    at __guard (http://localhost:8081/index.bundle//&platform=android&dev=true&minify=false&app=com.app.lv&modulesOnly=false&runModule=true:3281:15)
    at callFunctionReturnFlushedQueue (http://localhost:8081/index.bundle//&platform=android&dev=true&minify=false&app=com.app.lv&modulesOnly=false&runModule=true:3063:21)
 ERROR  [Error: Unhandled error. (Expected HTTP 101 response but was '401 Unauthorized')]
 LOG  Error has been symbolicated
Error: Unhandled error. (Expected HTTP 101 response but was '401 Unauthorized')
    at emit (/Users/pb/te/react-native-sdk/lv/node_modules/events/events.js:142:23)
    at _socket.onerror (/Users/pb/te/react-native-sdk/lv/node_modules/@deepgram/sdk/dist/main/packages/LiveClient.js:27:21)
 LOG  deepgram: keepalive
 LOG  deepgram: keepalive
 LOG  deepgram {"_events": {"open": [Function anonymous]}, "_eventsCount": 1, "_maxListeners": undefined, "_socket": {"CLOSED": 3, "CLOSING": 2, "CONNECTING": 0, "OPEN": 1, "_eventEmitter": {}, "_socketId": 3, "_subscriptions": [], "readyState": 3, "url": "wss://api.deepgram.com/v1/listen?language=en-US&detect_language=true&interim_results=true&punctuate=true&model=nova-2-general&smart_format=true"}, "baseUrl": "wss://api.deepgram.com/", "key": "API_KEY", "options": {"fetch": {"headers": [Object]}, "global": {"url": "api.deepgram.com"}}, "transcriptionOptions": {"detect_language": true, "interim_results": true, "language": "en-US", "model": "nova-2-general", "punctuate": true, "smart_format": true}}
```

Additional Information:
React Native Version: 0.72.6
Deepgram SDK Version: 3.1.6

Any help regarding this will be helpful.

Thanks

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by jkroll-deepgram** [comment]
Hi @DYSLEVIUM, can you please upgrade to JS-SDK version 3.1.7 or higher? This was a bug and we just released a fix today. Apologies for the inconvenience in your development.
https://github.com/deepgram/deepgram-js-sdk/releases/tag/v3.1.7

**Reply by DYSLEVIUM** [reply]
Hi @jkroll-deepgram, thanks for the prompt reply. I will test it using the version 3.1.7.

**Reply by DYSLEVIUM** [reply]
Hi @jkroll-deepgram, I am still getting the error after upgrading to 3.1.7 in my react native application(android).

```typescript
Error: Unhandled error. (Expected HTTP 101 response but was '401 Unauthorized')
    at emit (/Users/pb/te/react-native-sdk/lv/node_modules/events/events.js:142:23)
    at _socket.onerror (/Users/pb/te/react-native-sdk/lv/node_modules/@deepgram/sdk/dist/main/packages/LiveClient.js:27:21)
```
By the way, I am also using W3CWebSocket in some other part of my application and seems to work fine. Is there any specific issue with react native or the JS-SDK that might me causing this issue?

Let me know, if you need some other debug logs.

**Reply by jkroll-deepgram** [reply]
I'm sorry that this is still blocking you, I reopened the discussion.

Are you getting any Deepgram request IDs returned with these error messages?

As a sanity check that your API key is indeed valid, could you try a quick curl request? Here's a sample one that you can toss your own key into:
```
curl \
-X POST \
-H "Authorization: Token <key>" \
-H 'content-type: application/json' \
-d '{"url":"https://static.deepgram.com/examples/Bueller-Life-moves-pretty-fast.wav"}' \
"https://api.deepgram.com/v1/listen?smart_format=true&language=en&model=nova-2"
```

I'm not aware of any issues caused by using React native along with our SDK. I heard from a colleague that they are building a project with the JS SDK and React and haven't run into this issue, so I think we'll be able to move past this with a bit of debugging.

**Reply by DYSLEVIUM** [reply]
Hi @jkroll-deepgram, I tried using the curl request and it works as expected. But the error is still persisting for the JS-SDK.

However, I tried a different approach to get the live transcript working. Now, I was capturing the audio data from the react native application and sending it to a node server which made the request to the deepgram API using the deepgram JS-SDK. But I am getting a empty transcripts for all my attempts. I tried different encoding, sample rates; but nothing is working. I still get empty transcripts. Here are some of the request IDs:

1. 5fc12608-a789-4d06-bd52-e7e83ea35f59
1. cb46a6f3-03c1-454c-bdb4-65af2814d98e
1. 2dac8ec2-1b63-4f06-818e-35460ce65a9b
1. f1eddf85-3634-41a8-99d3-f658b3c5f8aa
1. 78d8319b-d535-4569-b471-49b8dab9226b

This is a sample response that I have been getting:

```typescript
{
  "type": "Results",
  "channel_index": [
    0,
    1
  ],
  "duration": 2.35,
  "start": 2.23,
  "is_final": true,
  "speech_final": true,
  "channel": {
    "alternatives": [
      {
        "transcript": "",
        "confidence": 0,
        "words": []
      }
    ]
  },
  "metadata": {
    "request_id": "50eb5991-cde4-4fb3-8ef5-870356f1e867",
    "model_info": {
      "name": "2-general-nova",
      "version": "2024-01-18.26916",
      "arch": "nova-2"
    },
    "model_uuid": "c0d1a568-ce81-4fea-97e7-bd45cb1fdf3c"
  }
}
```

**Reply by aagmangour** [reply]
I m facing the similar problem on my react native app, using deepgram js sdk v3.1.9.

deepgram connection: {
  "_events": {},
  "_eventsCount": 1,
  "key": "xxxxxxxxxxxxxxxxxxxxxxxxxx",
  "options": {
    "global": {
      "url": "https://api.deepgram.com"
    },
    "fetch": {
      "headers": {
        "Content-Type": "application/json",
        "X-Client-Info": "@deepgram/sdk; browser; v3.1.9",
        "User-Agent": "@deepgram/sdk/3.1.9"
      }
    }
  },
  "baseUrl": "wss://api.deepgram.com/",
  "transcriptionOptions": {
    "language": "hi",
    "interim_results": true,
    "punctuate": true,
    "model": "nova-2-general",
    "smart_format": true
  },
  "_socket": {
    "CONNECTING": 0,
    "OPEN": 1,
    "CLOSING": 2,
    "CLOSED": 3,
    "readyState": 0,
    "url": "wss://api.deepgram.com/v1/listen?language=hi&interim_results=true&punctuate=true&model=nova-2-general&smart_format=true",
    "_eventEmitter": {},
    "_socketId": 5,
    "_subscriptions": [
      {},
      {},
      {},
      {}
    ]
  }
}
 LOG  error: [Error: Unhandled error. (Expected HTTP 101 response but was '401 Unauthorized')] Error: Unhandled error. (Expected HTTP 101 response but was '401 Unauthorized')

**Comment by vijayjaybhay-bruviti** [comment]
> Hi @DYSLEVIUM, can you please upgrade to JS-SDK version 3.1.7 or higher? This was a bug and we just released a fix today. Apologies for the inconvenience in your development. https://github.com/deepgram/deepgram-js-sdk/releases/tag/v3.1.7

I am facing similar issue with v3.2.0 when used in react-native app. Any updates?

**Reply by vijayjaybhay-bruviti** [reply]
Finally I switched to custom websocket client for Live Transcription  in react native. You can use React Native's websocket class as follows
```
//No import needed, this is available in React Native as is
    this._socket = new WebSocket(url.toString(), null, {
      headers: {
        Authorization: 'Token ' + this.key,
      },
    });

```
You can replicate LiveClient.ts available in deepgram JS SDK(v3.20) as done here https://gist.github.com/vijayjaybhay-bruviti/b652572323c373d2e60ab3885ce94a12  for quick start or implement all callbacks from socket like CloseStream, KeepAlive events etc.  on your code  (You can also use patch in RN, my case was different so I didn't)

**Comment by ahaseeb001** [comment]
I am getting this error : "Invalid base URL: wss://api.deepgram.com" at deepgram?.listen?.live
i am using 
"@deepgram/sdk": "^3.5.0",
 "react": "18.1.0",
"react-native": "0.70.0",

```
  import { createClient } from "@deepgram/sdk";
  const CreateDeepGramClient = async () => {
    try {
      const deepgram = new createClient(DEEPGRAM_API_KEY);

      const live = await deepgram?.listen?.live({
        language: "en-US",
        detect_language: true,
        interim_results: true,
        punctuate: true,
        model: "nova-2-general",
        smart_format: true,
      });
      console.log("🚀 ~ useCallback ~ live:", live);
    } catch (error) {
      console.log("🚀 ~ CreateDeepGramClient ~ error:", error);
    }
  };
```

**Comment by montytorr** [comment]
if we follow the example here https://github.com/deepgram-devs/node-live-example/blob/main/server.js we still get "Unexpected server response: 401"

I've tried all possible ways, I'm sure of my key. using the SDK I only get 401 on deepgramClient.listen.live

**Reply by yoguib** [reply]
same here!

**Comment by jpvajda** [comment]
Hi Everyone, @lukeocodes  or @naomi-lgbt  might have some ideas here, I'll add them to this Discussion.

**Comment by lukeocodes** [comment, marked as answer]
This is a run away thread. 401 errors are an API unauthorized error. Thread locked.

The original thread was related to version 3.1.x and we're now at 3.5.x - given the hundreds of thousands of invocations between these versions it's safe to say that the SDK is not causing your 401 errors. My suggestion would be to create a new API key with no expiry and give that a go 

However, if you believe you've found a bug in the SDK, please raise it on the SDK repository with details about the version you're using, what the runtime is (browser? Node? Bun?) and a copy of the code you're using (minus the API key)
